README.md
# optical-keyboard

Keyboard PCB for Gateron optical switches (KiCad project).

- See [Optical Keyboard Design](https://girishji.github.io/jekyll/update/2022/08/17/optical-keyboard-design.html)
- There is a python script that is used for placing switches and laying out
  tracks. See also [KiCad 6 Python Scripting](https://girishji.github.io/)
- [QMK](https://github.com/girishji/qmk_firmware/tree/giri-optical-matrix/keyboards/opticalkb) code for this keyboard
- There is KiCad project for switch cutout plate under `switch-coutout` directory
- Part numbers for JLCPCB are embedded into symbols
- There is also a [single switch PCB](https://github.com/girishji/optical-amoeba)
- Another optical [keyboard](https://github.com/Dachtire/sok42) that might interest you
- Blackpill reference, including [pins to avoid](https://docs.qmk.fm/#/platformdev_blackpill_f411)

***


![image](https://i.imgur.com/hNEdrSI.jpg)
![image](https://i.imgur.com/ZMeAAxd.jpg)
![image](https://i.imgur.com/QnQpo4G.jpg)


ChangeLog
Author: Girish Palya <girishji@gmail.com>
Date:   Fri, 19 Aug 2022 17:10:22 +0200

update readme



M  README.md
M  switch-cutout/switch-cutout.kicad_prl

commit 8edb23d13ca5f3c54de6698be852a0eb747a6b07
Author: Girish Palya <girishji@gmail.com>
Date:   Wed Aug 3 11:14:24 2022 +0200

    Fix footprint hole size
    
    Cutouts should be 14x14 mm, not 13.8mm square; This fits
    both gateron and keychron optical switches
    
    M  Optical_Parts.pretty/Switch_Cutout.kicad_mod
    M  Optical_Parts.pretty/Switch_Cutout_1.25u.kicad_mod
    M  Optical_Parts.pretty/Switch_Cutout_1.5u.kicad_mod
    M  Optical_Parts.pretty/Switch_Cutout_1.75u.kicad_mod
    M  switch-cutout/fp-lib-table
    M  switch-cutout/switch-cutout.kicad_pcb
    M  switch-cutout/sym-lib-table

commit 70191034f0eb0216f1f753f3880bdcb7de0916d5
Author: Girish Palya <girishji@gmail.com>
Date:   Mon Aug 1 13:08:47 2022 +0200

    Add footprints library
    
    A  OpticalSymbols_Library.kicad_sym
    A  Optical_Parts.pretty/BlackPill_STLINK.kicad_mod
    A  Optical_Parts.pretty/BlackPill_STLINK2.kicad_mod
    A  Optical_Parts.pretty/Everlight_IR_IR12-21C_TR8.kicad_mod
    A  Optical_Parts.pretty/Everlight_PT_PT12-21C_TR8.kicad_mod
    A  Optical_Parts.pretty/Gateron_Optical_Switch.kicad_mod
    A  Optical_Parts.pretty/Gateron_Switch_Basic.kicad_mod
    A  Optical_Parts.pretty/MountingHole_1.2mm_M1.kicad_mod
    A  Optical_Parts.pretty/MountingHole_2.2mm_M2.kicad_mod
    A  Optical_Parts.pretty/MountingHole_2.2mm_M2_jlcpcb.kicad_mod
    A  Optical_Parts.pretty/Switch_Cutout.kicad_mod
    A  Optical_Parts.pretty/Switch_Cutout_1.25u.kicad_mod
    A  Optical_Parts.pretty/Switch_Cutout_1.5u.kicad_mod
    A  Optical_Parts.pretty/Switch_Cutout_1.75u.kicad_mod
    A  Optical_Parts.pretty/TestPoint_Plated_Hole_D1.2mm.kicad_mod
    A  Optical_Parts.pretty/TestPoint_THTPad_1.0x1.5mm_Drill0.5mm.kicad_mod
    A  Optical_Parts.pretty/TestPoint_THTPad_1.0x2.5mm_Drill0.5mm.kicad_mod
    M  README.md
    M  fp-lib-table
    M  optical-keyboard.kicad_pcb
    M  switch.kicad_sch
    M  sym-lib-table

commit c031835a9aca3336b69cdba5c9820a2adaa20fad
Author: Girish Palya <girishji@gmail.com>
Date:   Mon Aug 1 12:36:15 2022 +0200

    Add separate power source for PTs
    
    Sharing input pin b/w pt and led requires a diode, and diode
    slows down the response time of pt considerably
    
    M  optical-keyboard.kicad_pcb
    M  optical-keyboard.kicad_prl
    M  optical-keyboard.kicad_sch
    M  placefp.py
    M  switch.kicad_sch

commit 14a631e50e5605ef7dd8884ff0e848d931b2cee0
Author: Girish Palya <girishji@gmail.com>
Date:   Sun Jul 31 10:36:07 2022 +0200

    Update
    
    M  README.md

commit a35dcd8455f1d875e8cb3c01e79f531a5a159c2b
Author: Girish Palya <girishji@gmail.com>
Date:   Sun Jul 31 10:23:51 2022 +0200

    Fixed pic
    
    M  README.md

commit ed04fa910c0d7683aeec331894bd5e91ed50f8fd
Author: Girish Palya <girishji@gmail.com>
Date:   Sun Jul 31 10:20:54 2022 +0200

    Fixed bug in symbols by removing arrows
    
    M  optical-keyboard.kicad_sch
    M  switch.kicad_sch

commit 656b375c88896241ad135e091388fd7d2e6f9f9e
Author: Girish Palya <girishji@gmail.com>
Date:   Sat Jul 30 19:48:42 2022 +0200

    Added description
    
    M  README.md
    M  optical-keyboard.kicad_pro
    M  switch-cutout/switch-cutout.kicad_pro
    A  switch-cutout/sym-lib-table

commit e86094734db089b5fd0db38246ac00e46ae6fbe7
Author: Girish Palya <girishji@gmail.com>
Date:   Thu Jul 21 12:17:14 2022 +0200

    update footprint
    
    M  optical-keyboard.kicad_pcb

commit 1a6965cc3bf06c91457a46a45d71f9516bfe2bd4
Author: Girish Palya <girishji@gmail.com>
Date:   Wed Jul 20 09:55:38 2022 +0200

    update schematic
    
    M  optical-keyboard.kicad_sch
    M  switch.kicad_sch

commit d5249820a5e1c205f9ab33c431258aee1d90c852
Author: Girish Palya <girishji@gmail.com>
Date:   Wed Jul 20 09:52:31 2022 +0200

    updated pcb
    
    M  optical-keyboard.kicad_pcb
    M  optical-keyboard.kicad_pro
    M  optical-keyboard.kicad_sch
    M  placefp.py
    M  switch.kicad_sch

commit bf765a4ca861ad561dff2150cb606a80174742e6
Author: Girish Palya <girishji@gmail.com>
Date:   Tue Jul 19 23:50:18 2022 +0200

    first commit
    
    A  .gitignore
    A  fp-lib-table
    A  optical-keyboard.kicad_pcb
    A  optical-keyboard.kicad_prl
    A  optical-keyboard.kicad_pro
    A  optical-keyboard.kicad_sch
    A  placefp.py
    A  switch-cutout/.gitignore
    A  switch-cutout/fp-lib-table
    A  switch-cutout/switch-cutout.kicad_pcb
    A  switch-cutout/switch-cutout.kicad_prl
    A  switch-cutout/switch-cutout.kicad_pro
    A  switch-cutout/switch-cutout.kicad_sch
    A  switch.kicad_sch
    A  sym-lib-table

commit a43a9b83fdb1862fc516ac76f60818d7b1216632
Author: girishp <girishji@gmail.com>
Date:   Tue Jul 19 23:45:11 2022 +0200

    Initial commit


placefp.py
# Place footprints console script
#
# To run as script in python console,
#   place or symplink this script to ~/Documents/KiCad/6.0/scripting/plugins
#   Run from python console using 'import placefp'
#   To reapply:
#     import importlib
#     importlib.reload(placefp)
#  OR
#    exec(open("path-to-script-file").read())

import pcbnew
import math
import itertools


class Switch:
    # S, D, Q, RL
    _pos = {
        "S": (0, 0),
        "D": (-3.4, 0),
        "Q": (3.4, 0),
        "RL": (8.3, 4),
    }

    def __init__(self, board, num) -> None:
        if not board and not num:
            return  # dummy
        self.footprints = {
            "S": board.FindFootprintByReference("S" + str(num)),
            "D": board.FindFootprintByReference("D" + str(num)),
            "Q": board.FindFootprintByReference("Q" + str(num)),
            "RL": board.FindFootprintByReference("RL" + str(num)),
        }
        self.orient()

    def orient(self):
        self.footprints["S"].SetOrientation(0)  # 1/10 of degree
        self.footprints["D"].SetOrientation(-90 * 10)  # 1/10 of degree
        self.footprints["Q"].SetOrientation(90 * 10)  # 1/10 of degree
        self.footprints["RL"].SetOrientation(-90 * 10)  # 1/10 of degree

    @staticmethod
    def add_track(start, end, layer=pcbnew.F_Cu):
        board = pcbnew.GetBoard()
        track = pcbnew.PCB_TRACK(board)
        track.SetStart(start)
        track.SetEnd(end)
        track.SetWidth(int(0.3 * 1e6))
        track.SetLayer(layer)
        board.Add(track)

    def get_pad_point(self, fp, pad_num):
        return self.footprints[fp].FindPadByNumber(str(pad_num)).GetCenter()

    def get_track_end(self, fp, pad_num, dist, orient=0):
        start = self.get_pad_point(fp, pad_num)
        deg = self.footprints[fp].GetOrientation() // 10
        d = dist * 1e6  # mm
        deg = 90 + deg + orient
        return pcbnew.wxPoint(
            start.x + d * math.cos(math.radians(deg)),
            start.y - d * math.sin(math.radians(deg)),
        )

    def add_tracks(self):
        sta = self.get_pad_point("RL", 2)
        end = self.get_track_end("RL", 2, -8.5)
        Switch.add_track(sta, end)
        end2 = self.get_pad_point("D", 2)
        Switch.add_track(end, end2)

    def place(self, offset):
        for fp in self._pos.keys():
            p = pcbnew.wxPointMM(
                Switch._pos[fp][0] + offset[0], Switch._pos[fp][1] + offset[1]
            )
            self.footprints[fp].SetPosition(p)

    @staticmethod
    def place_fp(offset, fp, pos, orient):
        fp.SetOrientation(orient * 10)  # 1/10 of degree
        p = pcbnew.wxPointMM(pos[0] + offset[0], pos[1] + offset[1])
        fp.SetPosition(p)

    def rotate(self, deg):
        p = self.footprints["S"].GetPosition()
        for _, fp in self.footprints.items():
            fp.Rotate(p, deg * 10)


class Keyboard(object):
    def __init__(self) -> None:
        self.switches = [Switch(None, None)]
        board = pcbnew.GetBoard()
        for i in range(1, 76):
            self.switches.append(Switch(board, i))
        self.DIM = 19.05
        self.RP = [
            board.FindFootprintByReference("RP" + str(fp))
            for fp in [1] + list(range(1, 6))
        ]

    def place_footprints(self):
        dim = self.DIM
        board = pcbnew.GetBoard()
        rp_pos = (8.3, -0.5)

        # row 1
        for i in range(1, 16):
            self.switches[i].place((i * dim, 0))
        Switch.place_fp((4 * dim, 0), self.RP[1], rp_pos, 90)

        # row 2
        offs = dim + dim / 4
        self.switches[16].place((offs, dim))
        for i in range(17, 29):
            self.switches[i].place((offs + dim / 4 + (i - 16) * dim, dim))
        self.switches[29].place((offs + dim / 4 + dim * 13 + dim / 4, dim))
        Switch.place_fp((offs + dim / 4 + 3 * dim, dim), self.RP[2], rp_pos, 90)

        # row 3
        self.switches[31].place((dim / 2, 2 * dim))
        self.switches[32].place((3 * dim / 2 + dim / 8, 2 * dim))
        offs = 3 * dim / 2 + dim / 4
        for i in range(33, 44):
            self.switches[i].place((offs + (i - 32) * dim, 2 * dim))
        Switch.place_fp((offs + 2 * dim, 2 * dim), self.RP[3], rp_pos, 90)
        offs += 12 * dim + dim / 4
        self.switches[44].place((offs, 2 * dim))
        self.switches[45].place((offs + dim + dim / 4 + dim / 8, 2 * dim))

        # row 4
        offs = dim * (0.5 + 0.75 / 2)
        self.switches[46].place((offs, 3 * dim))
        offs += dim + dim * 0.75 / 2
        for i in range(47, 58):
            self.switches[i].place((offs + (i - 47) * dim, 3 * dim))
        Switch.place_fp((offs + 2 * dim, 3 * dim), self.RP[4], rp_pos, 90)
        offs += dim * 11
        self.switches[58].place((offs + dim / 8, 3 * dim))
        offs += dim * 1.25
        self.switches[59].place((offs, 3 * dim))
        self.switches[60].place((offs + dim, 3 * dim))

        # row 5
        for i in range(61, 64):
            self.switches[i].place((dim / 2 + (i - 61) * dim, 4 * dim))
        offs = dim / 2 + dim * 3
        self.switches[64].place((offs + dim / 8, 4 * dim))
        Switch.place_fp((offs + dim / 8, 4 * dim), self.RP[5], rp_pos, 90)
        self.switches[65].place((offs + dim * 1.25 + dim / 8, 4 * dim))
        offs += dim * 1.25 * 2
        self.switches[66].place((offs, 4.5 * dim))
        self.switches[66].rotate(-15 + 90)
        offs += dim * 1.25
        self.switches[67].place((offs, 4 * dim))
        self.switches[30].place((offs, 5 * dim))
        self.switches[68].place((offs + dim + dim / 4, 4.5 * dim))
        self.switches[68].rotate(15 + 90)
        offs += dim * 1.25
        for i in range(69, 76):
            self.switches[i].place((offs + (i - 68) * dim, 4 * dim))

        tp1 = board.FindFootprintByReference("TP1")
        tp2 = board.FindFootprintByReference("TP2")
        tp1.SetPosition(pcbnew.wxPointMM(dim * 6.0, dim * 3.7))
        tp2.SetPosition(pcbnew.wxPointMM(dim * 6.2, dim * 3.7))

        bp = board.FindFootprintByReference("U1")
        bp.SetOrientation(90 * 10)  # 1/10 of degree
        bp.SetPosition(pcbnew.wxPointMM(dim * 1.29, dim))

        pcbnew.Refresh()

    def remove_tracks(self):
        # delete tracks and vias
        board = pcbnew.GetBoard()
        tracks = board.GetTracks()
        for t in tracks:
            board.Delete(t)

    def add_via(self, loc):
        board = pcbnew.GetBoard()
        via = pcbnew.PCB_VIA(board)
        via.SetPosition(loc)
        via.SetDrill(int(0.3 * 1e6))
        via.SetWidth(int(0.6 * 1e6))
        board.Add(via)

    def _via_track(self, sta, end):
        Switch.add_track(sta, end)
        self.add_via(end)
        return end

    def via_track(self, sw, fp, pad_num, offset=-1.7):
        sta = sw.get_pad_point(fp, pad_num)
        end = sw.get_track_end(fp, pad_num, offset)
        return self._via_track(sta, end)

    def add_tracks(self):
        # add tracks
        for i in range(1, 76):
            self.switches[i].add_tracks()

        # rows
        for i in itertools.chain(
            range(1, 15),
            range(16, 29),
            range(31, 45),
            range(46, 60),
            range(61, 65),
            range(69, 75),
        ):
            sw1 = self.switches[i]
            start = sw1.get_pad_point("Q", 1)
            d = 1.5 * 1e6  # mm
            end = pcbnew.wxPoint(start.x + d, start.y - d)
            Switch.add_track(start, end)

            sw2 = self.switches[i + 1]
            start = sw2.get_pad_point("Q", 1)
            end2 = pcbnew.wxPoint(start.x - d, start.y - d)
            Switch.add_track(start, end2)
            Switch.add_track(end, end2)

        # columns
        exclude = [1, 2, 16, 15, 30, 66, 68]
        for i1, i2, i3, i4, i5 in zip(
            range(1, 16),
            range(16, 31),
            range(31, 46),
            range(46, 61),
            range(61, 76),
        ):
            vias = {
                sw: self.via_track(self.switches[sw], "RL", 1)
                for sw in [i1, i2, i3, i4, i5]
                if sw not in exclude
            }
            for st, en in [(i1, i2), (i2, i3), (i3, i4), (i4, i5)]:
                if st in exclude or en in exclude:
                    continue
                sta = vias[st]
                end = pcbnew.wxPoint(sta.x, sta.y + 4 * 1e6)
                Switch.add_track(sta, end, pcbnew.B_Cu)
                sta = vias[en]
                end2 = pcbnew.wxPoint(sta.x, sta.y - 7 * 1e6)
                Switch.add_track(sta, end2, pcbnew.B_Cu)
                Switch.add_track(end, end2, pcbnew.B_Cu)

        # RPs
        vias = {}
        for rp in range(1, 6):
            start = self.RP[rp].FindPadByNumber(str(2)).GetCenter()
            d = 1.7 * 1e6
            end = pcbnew.wxPoint(start.x + d, start.y)
            vias[rp] = self._via_track(start, end)
        for st, en in zip(range(1, 5), range(2, 6)):
            sta = vias[st]
            end = pcbnew.wxPoint(sta.x, sta.y + 9 * 1e6)
            Switch.add_track(sta, end, pcbnew.B_Cu)
            sta = vias[en]
            end2 = pcbnew.wxPoint(sta.x, sta.y - 2 * 1e6)
            Switch.add_track(sta, end2, pcbnew.B_Cu)
            Switch.add_track(end, end2, pcbnew.B_Cu)

        # ground
        for i in range(1, 76):
            sta = self.switches[i].get_pad_point("Q", 2)
            end = self.switches[i].get_track_end("Q", 2, -1.5)
            Switch.add_track(sta, end)
            self.add_via(end)
            sta = self.switches[i].get_pad_point("D", 1)
            end = self.switches[i].get_track_end("D", 1, -1.5)
            Switch.add_track(sta, end)
            self.add_via(end)

        pcbnew.Refresh()

    def add_holes(self):
        dim = self.DIM
        board = pcbnew.GetBoard()
        holes = [board.FindFootprintByReference("H1")]  # dummy
        holes += [
            board.FindFootprintByReference("H" + str(num)) for num in range(1, 21)
        ]
        holes[1].SetPosition(pcbnew.wxPointMM(dim * 3, dim * 0.5))
        holes[2].SetPosition(pcbnew.wxPointMM(dim * 4, dim * 1.5))
        holes[3].SetPosition(pcbnew.wxPointMM(dim * (1 + 1 / 8), dim * 2.0))
        holes[4].SetPosition(pcbnew.wxPointMM(dim * (1 / 4 + 1 / 8), dim * 3.0))
        holes[5].SetPosition(pcbnew.wxPointMM(dim * (2 + 1 / 2 + 1 / 4), dim * 3.5))
        holes[6].SetPosition(pcbnew.wxPointMM(dim * (4 + 3 / 4), dim * 2.5))
        holes[7].SetPosition(pcbnew.wxPointMM(dim * (6 + 3 / 4), dim * 2.5))
        holes[8].SetPosition(
            pcbnew.wxPointMM(dim * (5 + 3 / 4 - 1 / 8), dim * (3 + 1 / 2 + 1 / 8))
        )
        holes[9].SetPosition(
            pcbnew.wxPointMM(dim * (8 + 3 / 4 + 1 / 8), dim * (3 + 1 / 2 + 1 / 8))
        )
        holes[10].SetPosition(pcbnew.wxPointMM(dim * 8, dim * 0.5))
        holes[11].SetPosition(pcbnew.wxPointMM(dim * 10, dim * 1.5))
        holes[12].SetPosition(pcbnew.wxPointMM(dim * 12, dim * 1.5))
        holes[13].SetPosition(
            pcbnew.wxPointMM(dim * (14 + 1 / 8), dim * (1 / 2 + 1 / 8))
        )
        holes[14].SetPosition(pcbnew.wxPointMM(dim * (14 + 1 / 2 + 1 / 4), dim * (2)))
        holes[15].SetPosition(pcbnew.wxPointMM(dim * (12), dim * (3 + 1 / 2)))
        holes[16].SetPosition(pcbnew.wxPointMM(dim * (14), dim * (3 + 1 / 2 - 1 / 8)))

        holes[17].SetPosition(pcbnew.wxPointMM(dim * (3 + 1 / 2 + 1 / 4), dim * 3.5))
        holes[18].SetPosition(pcbnew.wxPointMM(dim * 14, dim * 1.5))
        holes[19].SetPosition(pcbnew.wxPointMM(dim * (11 - 1 / 4), dim * (2 + 1 / 2)))
        holes[20].SetPosition(pcbnew.wxPointMM(dim * (1 / 4 + 1 / 8), dim * 2.5))

        pcbnew.Refresh()


kb = Keyboard()
kb.place_footprints()
kb.remove_tracks()
kb.add_tracks()
kb.add_holes()
